File: H4x2-TinySDK/H4x2-TinySDK/Ed25519/Point.cs
// 
// Tide Protocol - Infrastructure for a TRUE Zero-Trust paradigm
// Copyright (C) 2022 Tide Foundation Ltd
// 
// This program is free software and is subject to the terms of 
// the Tide Community Open Code License as published by the 
// Tide Foundation Limited. You may modify it and redistribute 
// it in accordance with and subject to the terms of that License.
// This program is distributed WITHOUT WARRANTY of any kind, 
// including without any implied warranty of MERCHANTABILITY or 
// FITNESS FOR A PARTICULAR PURPOSE.
// See the Tide Community Open Code License for more details.
// You should have received a copy of the Tide Community Open 
// Code License along with this program.
// If not, see https://tide.org/licenses_tcoc2-0-0-en
//

using System.Numerics;
using System.Text;
using H4x2_TinySDK.Tools;

namespace H4x2_TinySDK.Ed25519
{
    /// <summary>
    /// Represents a point on the Ed25519 Curve.
    /// </summary>
    public class Point
    {
        private BigInteger X { get; }
        private BigInteger Y { get; }
        private BigInteger Z { get; }
        private BigInteger T { get; }

        /// <summary>
        /// Create a point from extended coordinates. Consider passing only x and y for simpler use.
        /// </summary>
        /// <param name="x"></param>
        /// <param name="y"></param>
        /// <param name="z"></param>
        /// <param name="t"></param>
        public Point(BigInteger x, BigInteger y, BigInteger z, BigInteger t)
        {
            X = x;
            Y = y;
            Z = z;
            T = t;
        }
        /// <summary>
        /// Create a point from normal coordinates.
        /// </summary>
        /// <param name="x"></param>
        /// <param name="y"></param>
        public Point(BigInteger x, BigInteger y)
        {
            X = x;
            Y = y;
            Z = 1;
            T = Mod(x * y);
        }
        /// <summary>
        /// Creates a point from bytes.
        /// </summary>
        /// <param name="data"></param>
        /// <param name="compact"></param>
        /// <returns></returns>
        public static Point FromBytes(IReadOnlyList<byte> data)
        {
            var x = new BigInteger(data.Take(32).ToArray(), true, false);
            var y = new BigInteger(data.Skip(32).ToArray(), true, false);
            var point = new Point(x, y);
            return point;
        }
        public static Point FromBase64(string data) => Point.FromBytes(Convert.FromBase64String(data));
        /// <summary>
        /// Performs ( X * modular_inverse(Z) ) % M to get the actual x coordinate.
        /// </summary>
        /// <returns>The actual x coordinate of this point.</returns>
        public BigInteger GetX() => Mod(X * BigInteger.ModPow(Z, Curve.M - 2, Curve.M));
        /// <summary>
        /// Performs ( Y * modular_inverse(Z) ) % M to get the actual y coordinate.
        /// </summary>
        /// <returns>The actual y coordinate of this point.</returns>
        public BigInteger GetY() => Mod(Y * BigInteger.ModPow(Z, Curve.M - 2, Curve.M));
        /// <summary>
        /// Determines if two points are equal to each other.
        /// </summary>
        /// <param name="other"></param>
        /// <returns></returns>
        public bool isEqual(Point other)
        {
            BigInteger X1Z2 = Mod(this.X * other.Z);
            BigInteger X2Z1 = Mod(other.X * this.Z);
            BigInteger Y1Z2 = Mod(this.Y * other.Z);
            BigInteger Y2Z1 = Mod(other.Y * this.Z);
            return (X1Z2 == X2Z1) && (Y1Z2 == Y2Z1);
        }
        /// <summary>
        /// Determines whether this point is a valid point on the Ed25519 Curve.
        /// </summary>
        /// <returns>A boolean whether the point it is or isn't on the curve.</returns>
        public bool IsValid()
        {
            BigInteger y = GetY();
            BigInteger x = GetX();
            BigInteger y2 = Mod(y * y);
            BigInteger x2 = Mod(x * x);
            return Mod(y2 * Mod(1 + Curve.Not_Minus_D * x2)) == Mod(1 + x2);
        }
        /// <summary>
        /// Determines if this point is safe for point multiplication.
        /// </summary>
        /// <returns></returns>
        public bool IsSafePoint()
        {
            if (this.IsInfinity())
                return false;
            /*This check will be always pass since we use %M*/
            if (this.GetX() < 0 || this.GetY() < 0 || this.GetX() >= Curve.M || this.GetY() >= Curve.M)
                return false;
            if (!this.IsValid())
                return false;
            if (!(this * Curve.N).IsInfinity())
                return false;
            return true;
        }
        /// <summary>
        /// Determines if this point is the infinity point.
        /// </summary>
        /// <returns></returns>
        public bool IsInfinity()
        {
            return this.isEqual(Curve.Infinity);
        }
        /// <summary>
        /// </summary>
        /// <returns>The point coordinates as unsigned, little endian byte arrays. 
        public byte[] ToByteArray()
        {
            return this.GetX().ToByteArray(true, false).PadRight(32)
                    .Concat(this.GetY().ToByteArray(true, false).PadRight(32)).ToArray();
        }
        /// <summary>
        /// </summary>
        /// <returns>The point as a base64 encoded string</returns>
        public string ToBase64()
        {
            return Convert.ToBase64String(this.ToByteArray());
        }
        /// <summary>
        /// Compresses the point into a 32 byte array.
        /// </summary>
        /// <returns></returns>
        public byte[] Compress()
        {
            var x_lsb = this.GetX() & 1;
            // Encode the y-coordinate as a little-endian string of 32 octets.
            byte[] yByteArray = this.GetY().ToByteArray(true, false).PadRight(32);
            int new_msb = 0;
            if (x_lsb == 1)
            {
                new_msb = yByteArray[31] | 128;
            }
            if (x_lsb == 0)
            {
                new_msb = yByteArray[31] & 127;
            }
            // Copy the least significant bit of the x - coordinate to the most significant bit of the final octet.
            yByteArray[31] = (byte)new_msb;
            return yByteArray;
        }
        /// <summary>
        /// Multiplies a point by a scalar using double and add algorithm on the Ed25519 Curve.
        /// Does not perform safety checks on scalar or the point, yet.
        /// </summary>
        /// <param name="point"></param>
        /// <param name="num"></param>
        /// <returns>A new point on the Ed25519 Curve.</returns>
        public static Point operator *(Point point, BigInteger num)
        {
            Point newPoint = new Point(BigInteger.Zero, BigInteger.One, BigInteger.One, BigInteger.Zero);
            while (num > BigInteger.Zero)
            {
                if ((num & BigInteger.One).Equals(BigInteger.One)) newPoint = newPoint + point;
                point = Double(point);
                num = num >> 1;
            }
            return newPoint;
        }
        /// <summary>
        /// Add a point by itself ("double") on the Ed25519 Curve.
        /// </summary>
        /// <param name="point"></param>
        /// <returns>A new point on the Ed25519 Curve.</returns>
        public static Point Double(in Point point)
        {
            // Algorithm taken from https://www.hyperelliptic.org/EFD/g1p/auto-twisted-extended-1.html.

            BigInteger A = Mod(point.X * point.X);
            BigInteger B = Mod(point.Y * point.Y);
            BigInteger C = Mod(Curve.Two * Mod(point.Z * point.Z));
            BigInteger D = Mod(Curve.A * A);
            BigInteger x1y1 = point.X + point.Y;
            BigInteger E = Mod(Mod(x1y1 * x1y1) - A - B);
            BigInteger G = D + B;
            BigInteger F = G - C;
            BigInteger H = D - B;
            BigInteger X3 = Mod(E * F);
            BigInteger Y3 = Mod(G * H);
            BigInteger T3 = Mod(E * H);
            BigInteger Z3 = Mod(F * G);
            return new Point(X3, Y3, Z3, T3);
        }
        /// <summary>
        /// Adds two points on the Ed25519 Curve. Currently, does not check if point is on curve or prime order group.
        /// </summary>
        /// <param name="point1"></param>
        /// <param name="point2"></param>
        /// <returns>A new point on the Ed25519 Curve.</returns>
        public static Point operator +(in Point point1, in Point point2)
        {
            // TODO: check to see if point is on curve and on the prime order subgroup
            // Algorithm taken from https://www.hyperelliptic.org/EFD/g1p/auto-twisted-extended-1.html.

            BigInteger A = Mod((point1.Y - point1.X) * (point2.Y + point2.X));
            BigInteger B = Mod((point1.Y + point1.X) * (point2.Y - point2.X));
            BigInteger F = Mod(B - A);
            if (F.Equals(BigInteger.Zero)) return Double(point1);
            BigInteger C = Mod(point1.Z * Curve.Two * point2.T);
            BigInteger D = Mod(point1.T * Curve.Two * point2.Z);
            BigInteger E = D + C;
            BigInteger G = B + A;
            BigInteger H = D - C;
            BigInteger X3 = Mod(E * F);
            BigInteger Y3 = Mod(G * H);
            BigInteger T3 = Mod(E * H);
            BigInteger Z3 = Mod(F * G);
            return new Point(X3, Y3, Z3, T3);
        }
        private static BigInteger Mod(BigInteger a)
        {
            BigInteger res = a % Curve.M;
            return res >= BigInteger.Zero ? res : Curve.M + res;
        }
    }
}

File: Tidenclave/modules/H4x2-TideJS/Ed25519/point.js
// 
// Tide Protocol - Infrastructure for a TRUE Zero-Trust paradigm
// Copyright (C) 2022 Tide Foundation Ltd
// 
// This program is free software and is subject to the terms of 
// the Tide Community Open Code License as published by the 
// Tide Foundation Limited. You may modify it and redistribute 
// it in accordance with and subject to the terms of that License.
// This program is distributed WITHOUT WARRANTY of any kind, 
// including without any implied warranty of MERCHANTABILITY or 
// FITNESS FOR A PARTICULAR PURPOSE.
// See the Tide Community Open Code License for more details.
// You should have received a copy of the Tide Community Open 
// Code License along with this program.
// If not, see https://tide.org/licenses_tcoc2-0-0-en
//


import { SHA256_Digest } from "../Tools/Hash.js";
import { BigIntToByteArray, BigIntFromByteArray, bytesToBase64, base64ToBytes, ConcatUint8Arrays, Bytes2Hex } from "../Tools/Utils.js";

const _0n = BigInt(0);
const _1n = BigInt(1);
const _2n = BigInt(2);

export default class Point {

    static get a() {return BigInt(-1)};
    static get d() {return BigInt("37095705934669439343138083508754565189542113879843219016388785533085940283555");}
    static get d_FOR_Y_FROM_X() {return BigInt("20800338683988658368647408995589388737092878452977063003340006470870624536394");}
    static get p() {return BigInt("57896044618658097711785492504343953926634992332820282019728792003956564819949");}
    static get order() { return BigInt("7237005577332262213973186563042994240857116359379907606001950938285454250989");}
    static get g() { return new Point(
        BigInt("15112221349535400772501151409588531511454012693041857206046113283949847762202"),
        BigInt("46316835694926478169428394003475163141307993866256225615783033603165251855960"), 
        _1n, 
        BigInt("46827403850823179245072216630277197565144205554125654976674165829533817101731")
    );}
    static get infinity() {return new Point(_0n, _1n, _1n, _0n);} //infinity also known as identity for ed25519
    static get SQRT_M1() {return BigInt("19681161376707505956807079304988542015446066515923890162744021073123829784752");}


    /**
     * @param {bigint} x 
     * @param {bigint} y 
     * @param {bigint} z 
     * @param {bigint} t 
     */
    constructor(x, y=null, z=null, t=null) {
        this.x = x;
        this.y = y !== null ? y : getYfromX(x);
        this.z = z !== null ? z : _1n;
        this.t = t !== null ? t : mod(this.x * this.y);
    }
    
    isInfinity(){
        return this.isEqual(Point.infinity); 
    }

    isEqual(other){
        const X1Z2 = mod(this.x * other.z);
        const X2Z1 = mod(other.x * this.z);
        const Y1Z2 = mod(this.y * other.z);
        const Y2Z1 = mod(other.y * this.z);
        return X1Z2 === X2Z1 && Y1Z2 === Y2Z1;
    }

    negate(){
        return new Point(mod(-this.x), this.y, this.z, mod(-this.t)); //// not checked if works
    }


    /**
     * 
     * @param {bigint} num 
     * @returns {Point}
     */
    times(num) {
        var point = new Point(this.x, this.y, this.z, this.t);
        let newPoint = Point.infinity;
        while (num > _0n) {
            if ((num & _1n) === (_1n)) {
                newPoint = newPoint.add(point);
            }
            point = point.double();
            num = num >> _1n;
        }
        return newPoint;
    }

    /**
     * 
     * @returns {Point}
     */
    double() {
        let A = mod(this.x * this.x);
        let B = mod(this.y * this.y);
        let C = mod(_2n * mod(this.z * this.z));
        let D = mod(Point.a * A);
        let x1y1 = this.x + this.y;
        let E = mod(mod(x1y1 * x1y1) - A - B);
        let G = D + B;
        let F = G - C;
        let H = D - B;
        let X3 = mod(E * F);
        let Y3 = mod(G * H);
        let T3 = mod(E * H);
        let Z3 = mod(F * G);
        return new Point(X3, Y3, Z3, T3);
    }

    /**
     * @param {Point} other 
     * @returns {Point}
     */
    add(other) {
        let A = mod((this.y - this.x) * (other.y + other.x));
        let B = mod((this.y + this.x) * (other.y - other.x));
        let F = mod(B - A);
        if (F == _0n) return this.double();
        let C = mod(this.z * _2n * other.t);
        let D = mod(this.t * _2n * other.z);
        let E = D + C;
        let G = B + A;
        let H = D - C;
        let X3 = mod(E * F);
        let Y3 = mod(G * H);
        let T3 = mod(E * H);
        let Z3 = mod(F * G);
        return new Point(X3, Y3, Z3, T3);
    }

    /**
     * 
     * @returns {bigint}
     */
    getX(){
        return mod(this.x * mod_inv(this.z));
    }
    /**
     * 
     * @returns {bigint}
     */
    getY(){
        return mod(this.y * mod_inv(this.z));
    }
    /**
     * @param {Uint8Array} data
     * @returns {Point}
     */
    static from(data){
        var x = BigIntFromByteArray(data.slice(0, 32));
        var y = BigIntFromByteArray(data.slice(32, 64));
        return new Point(x, y);
    }
    /**
     * @param {string} data
     * @returns {Point}
     */
    static fromB64(data){
        return this.from(base64ToBytes(data));
    }

    static async fromString(message){
        // this function genereates a pseudo random point (multiple of G) from a message
        var x = BigIntFromByteArray(await SHA256_Digest(message));

        var point = new Point(x);

        while (!point.times(Point.order).isInfinity()) {
            x = x + _1n; 
            point = new Point(x);
        }

        return point;
    }

    /** @returns {Uint8Array} */
    toArray(){
        var xBuff = BigIntToByteArray(this.getX());
        var yBuff = BigIntToByteArray(this.getY());
        return ConcatUint8Arrays([xBuff, yBuff]);
    }

    /**
     * @returns {string}
     */
    toBase64(){
        return bytesToBase64(this.toArray());
    }
}

/**
 * @param {bigint} a 
 * @param {bigint} b 
 * @returns {bigint}
 */
function mod(a, b = Point.p) {
    var res = a % b;
    return res >= _0n ? res : b + res;
}

/**
 * @param {bigint} number 
 * @param {bigint} modulo 
 * @returns {bigint}
 */
function mod_inv(number, modulo = Point.p) {
    if (number === _0n || modulo <= _0n) {
        throw new Error(`invert: expected positive integers, got n=${number} mod=${modulo}`);
    }
    let a = mod(number, modulo);
    let b = modulo;
    // prettier-ignore
    let x = _0n, y = _1n, u = _1n, v = _0n;
    while (a !== _0n) {
        const q = b / a;
        const r = b % a;
        const m = x - u * q;
        const n = y - v * q;
        // prettier-ignore
        b = a, a = r, x = u, y = v, u = m, v = n;
    }
    const gcd = b;
    if (gcd !== _1n) throw new Error('invert: does not exist');
    return mod(x, modulo);
}

/**
 * @param {bigint} x 
 * @returns 
 */
 function getYfromX(x){
    // Always get 'higher' point. Remember there are two possible values, this always picks the first one to remain consistent.
    var x2 = mod(x * x);
    var u = mod(_1n + x2);
    var v = mod(_1n + Point.d_FOR_Y_FROM_X*x2);

    return uvRatio(u, v).value;
}

/**
 * @param {bigint} x 
 * @param {bigint} power 
 * @returns 
 */
function pow2(x, power){
    while (power-- > _0n) {
        x = mod(x * x);
    }
    return x;
}
// Power to (p-5)/8 aka x^(2^252-3)
// Used to calculate y - the square root of y².
// Exponentiates it to very big number.
// We are unwrapping the loop because it's 2x faster.
// (2n**252n-3n).toString(2) would produce bits [250x 1, 0, 1]
// We are multiplying it bit-by-bit
/**
 * 
 * @param {bigint} x 
 * @returns 
 */
//works
function pow_2_252_3(x) {
    const P = Point.p;
    const _5n = BigInt(5);
    const _10n = BigInt(10);
    const _20n = BigInt(20);
    const _40n = BigInt(40);
    const _80n = BigInt(80);
    const x2 = (x * x) % P;
    const b2 = (x2 * x) % P; // x^3, 11
    const b4 = (pow2(b2, _2n) * b2) % P; // x^15, 1111
    const b5 = (pow2(b4, _1n) * x) % P; // x^31
    const b10 = (pow2(b5, _5n) * b5) % P;
    const b20 = (pow2(b10, _10n) * b10) % P;
    const b40 = (pow2(b20, _20n) * b20) % P;
    const b80 = (pow2(b40, _40n) * b40) % P;
    const b160 = (pow2(b80, _80n) * b80) % P;
    const b240 = (pow2(b160, _80n) * b80) % P;
    const b250 = (pow2(b240, _10n) * b10) % P;
    const pow_p_5_8 = (pow2(b250, _2n) * x) % P;
    // ^ To pow to (p+3)/8, multiply it by x.
    return {pow_p_5_8, b2};
}
// this method is incomplete. not ready for ed25519 point decoding
/**
 * 
 * @param {bigint} u 
 * @param {bigint} v 
 * @returns 
 */
function uvRatio(u, v) {
    const v3 = mod(v * v * v);                  // v³
    const v7 = mod(v3 * v3 * v);                // v⁷
    const pow = pow_2_252_3(u * v7).pow_p_5_8;
    let x = mod(u * v3 * pow);                  // (uv³)(uv⁷)^(p-5)/8
    const vx2 = mod(v * x * x);                 // vx²
    const root1 = x;                            // First root candidate
    const root2 = mod(x * Point.SQRT_M1);             // Second root candidate
    const useRoot1 = vx2 === u;                 // If vx² = u (mod p), x is a square root
    const useRoot2 = vx2 === mod(-u);           // If vx² = -u, set x <-- x * 2^((p-1)/4)
    const noRoot = vx2 === mod((_0n-u) * Point.SQRT_M1);   // There is no valid root, vx² = -u√(-1)
    if (useRoot1) x = root1;
    if (useRoot2 || noRoot) x = root2;          // We return root2 anyway, for const-time
    if (edIsNegative(x)) x = mod(-x);
    return { isValid: useRoot1 || useRoot2, value: x };
  }
// Little-endian check for first LE bit (last BE bit);
function edIsNegative(num) {
    return (mod(num) & _1n) === _1n;
}